drivers/dht/dht.c

```diff
@@ -38,27 +38,55 @@
 #define ENABLE_DEBUG    (0)
 #include "debug.h"
 
+/* Every pulse send by the DHT longer than 40µs is interpreted as 1 */
 #define PULSE_WIDTH_THRESHOLD       (40U)
+/* If an expected pulse is not detected within 1000µs, something is wrong */
+#define TIMEOUT                     (1000U)
+/* The DHT sensor cannot measure more than once a second */
+#define DATA_HOLD_TIME              (1000U * US_PER_MS)
 
-static uint16_t read(gpio_t pin, int bits)
+static inline void _reset(dht_t *dev)
 {
+    gpio_init(dev->params.pin, GPIO_OUT);
+    gpio_set(dev->params.pin);
+
+}
+
+static int _read(uint16_t *dest, gpio_t pin, int bits)
+{
+    DEBUG("read\n");
     uint16_t res = 0;
 
     for (int i = 0; i < bits; i++) {
         uint32_t start, end;
         res <<= 1;
         /* measure the length between the next rising and falling flanks (the
          * time the pin is high - smoke up :-) */
-        while (!gpio_read(pin)) {}
+        unsigned counter = 0;
+        while (!gpio_read(pin)) {
```

Comment:
this is subjective, but I find this rather hard to read and understand. I would write this as:

```C
unsigned counter = TIMEOUT;
while (!gpio_read(pin) && counter--) {
    xtimer_usleep(1);
}
if (!counter) {
    return -1;
}
```

however, this is non-blocking

Reply:
Counting backwards could indeed be more intuitive. But I think counter would underflow that way, wouldn't it? E.g. it would break the loop when `counter` reaches zero, but than still does the post-decrement, right?

Reply:
mhm, right but also `unsigned` might be unsafe here anyway, i.e., when running on atmega arduinos (8bit?) so better use `int32_t` instead an thus avoid the underflow too

Reply:
I refactored all four instances of waiting for the GPIO to reach a specific logic level with a timeout into `static inline` function and changed the logic to count down to zero, as you suggested.

Reply:
OK, code-wise this seems to be the last issue open now. As I addressed this slightly different suggested: Can I hit the `resolve` button, or do you have further comments on how I addressed this instead?